Don't cancel context menu filtering when filling one of them fails

Let's make sure issue is not already fixed in latest builds first.

- [ ] I verified that Eclipse 2023-03 doesn't provide the feature. Pretty sure the later versions also don't

## Suggestion

From a fresh CDT 2023-03 installation and clean workspace:

* Create a c project with a `hello.c` file
* Install Sigasi Studio 5.3.3
* Right-click the file in the Project Explorer
* Because Sigasi Studio throws an error while creating its menu item, no filtering is applied to contribution items of other plugins, resulting in a very large context menu.

| correct | bugged |
| :-------: | :---------: |
| ![291613402-a49322a0-782b-4b32-b583-10da24dea88b](https://github.com/eclipse-platform/eclipse.platform.ui/assets/3389524/3ccbfb05-e77b-4ac3-9f43-f2b0616ef4aa) | ![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/3389524/ed1977af-4ee9-4c02-8c11-d48afd18c621) |

## Community

- [X] I understand suggesting an enhancement doesn't mandate anyone to implement it. Other contributors may consider this suggestion, or not, at their own convenience. The most efficient way to get it fixed is that I implement it myself and contribute it back as a good quality patch to the project.

## Stacktrace

We created a contribution item (probably ten years ago) that doesn't show itself at all (not even greyed out) when it shouldn't. In our latest release, a bug was introduced that caused it to throw a runtime exception, causing the above trouble. The fact that one plugin throws exceptions shouldn't destroy the experience for other plugins. We should gracefully handle the said exception.

Looking at it quickly, I think we can catch it either in `DynamicMenuContributionItem.fill()` or in [MenuManager.update()](https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/bundles/org.eclipse.jface/src/org/eclipse/jface/action/MenuManager.java#L805)

```
org.eclipse.e4.core.di.InjectionException: java.lang.NullPointerException
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:68)
	at org.eclipse.swt.widgets.Synchronizer.syncExec(Synchronizer.java:183)
	at org.eclipse.ui.internal.UISynchronizer.syncExec(UISynchronizer.java:133)
	at org.eclipse.swt.widgets.Display.syncExec(Display.java:5960)
	at org.eclipse.e4.ui.workbench.swt.DisplayUISynchronize.syncExec(DisplayUISynchronize.java:34)
	at org.eclipse.e4.ui.internal.di.UIEventObjectSupplier$UIEventHandler.handleEvent(UIEventObjectSupplier.java:64)
	at org.eclipse.equinox.internal.event.EventHandlerWrapper.handleEvent(EventHandlerWrapper.java:205)
	at org.eclipse.equinox.internal.event.EventHandlerTracker.dispatchEvent(EventHandlerTracker.java:203)
	at org.eclipse.equinox.internal.event.EventHandlerTracker.dispatchEvent(EventHandlerTracker.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151)
	at org.eclipse.equinox.internal.event.EventAdminImpl.dispatchEvent(EventAdminImpl.java:133)
	at org.eclipse.equinox.internal.event.EventAdminImpl.sendEvent(EventAdminImpl.java:75)
	at org.eclipse.equinox.internal.event.EventComponent.sendEvent(EventComponent.java:44)
	at org.eclipse.e4.ui.services.internal.events.EventBroker.send(EventBroker.java:55)
	at org.eclipse.e4.ui.internal.workbench.UIEventPublisher.notifyChanged(UIEventPublisher.java:63)
	at org.eclipse.emf.common.notify.impl.BasicNotifierImpl.eNotify(BasicNotifierImpl.java:424)
	at org.eclipse.emf.ecore.util.EcoreEList.dispatchNotification(EcoreEList.java:249)
	at org.eclipse.emf.common.notify.impl.NotifyingListImpl.addUnique(NotifyingListImpl.java:356)
	at org.eclipse.emf.common.util.AbstractEList.add(AbstractEList.java:340)
	at org.eclipse.e4.ui.workbench.renderers.swt.ContributionRecord.mergeIntoModel(ContributionRecord.java:253)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRenderer.processAddition(MenuManagerRenderer.java:551)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRenderer.generateContributions(MenuManagerRenderer.java:531)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRenderer.processContributions(MenuManagerRenderer.java:499)
	at org.eclipse.ui.internal.PopupMenuExtender.addMenuContributions(PopupMenuExtender.java:380)
	at org.eclipse.ui.internal.PopupMenuExtender.menuAboutToShow(PopupMenuExtender.java:355)
	at org.eclipse.jface.action.MenuManager.fireAboutToShow(MenuManager.java:338)
	at org.eclipse.jface.action.MenuManager.handleAboutToShow(MenuManager.java:468)
	at org.eclipse.jface.action.MenuManager$2.menuShown(MenuManager.java:495)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:259)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5855)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1529)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1555)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1534)
	at org.eclipse.swt.widgets.Menu._setVisible(Menu.java:290)
	at org.eclipse.swt.widgets.Display.runPopups(Display.java:5107)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4508)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:643)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:550)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
Caused by: java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:878)
	at com.sigasi.hdt.shared.ui.librarymapping.LibraryMappingExcludeContribution.fill(LibraryMappingExcludeContribution.java:4)
	at org.eclipse.ui.internal.menus.DynamicMenuContributionItem.fill(DynamicMenuContributionItem.java:146)
	at org.eclipse.jface.action.MenuManager.doItemFill(MenuManager.java:729)
	at org.eclipse.jface.action.MenuManager.update(MenuManager.java:806)
	at org.eclipse.jface.action.MenuManager.update(MenuManager.java:673)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRenderer.scheduleManagerUpdate(MenuManagerRenderer.java:1227)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRenderer.processContents(MenuManagerRenderer.java:655)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRenderer.subscribeTopicChildAdded(MenuManagerRenderer.java:325)
	at jdk.internal.reflect.GeneratedMethodAccessor31.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:58)
	... 59 more
```

--- a/bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/MenuManagerRenderer.java
+++ b/bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/MenuManagerRenderer.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2009, 2020 IBM Corporation and others.
+ * Copyright (c) 2009, 2023 IBM Corporation and others.
  *
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
@@ -22,6 +22,7 @@
  *     Daniel Kruegler <daniel.kruegler@gmail.com> - Bug 473779
  *     Simon Scholz <simon.scholz@vogella.com> - Bug 506306
  *     Axel Richard <axel.richard@oebo.fr> - Bug 354538
+ *     Christoph Läubrich - issue #1435
  *******************************************************************************/
 package org.eclipse.e4.ui.workbench.renderers.swt;
 
@@ -37,6 +38,7 @@ import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import org.eclipse.core.runtime.ILog;
 import org.eclipse.e4.core.commands.ExpressionContext;
 import org.eclipse.e4.core.contexts.ContextInjectionFactory;
 import org.eclipse.e4.core.contexts.IContextFunction;
@@ -1170,7 +1172,16 @@ public class MenuManagerRenderer extends SWTPartRenderer {
 							mgrToUpdate.clear();
 						}
 						for (IContributionManager mgr1 : toUpdate) {
-							mgr1.update(false);
+							try {
+								mgr1.update(false);
+							} catch (RuntimeException e) {
+								String message = String.format(
+										"ContributionManager '%s' threw an exception while performing update!", mgr1); //$NON-NLS-1$
+								ILog.get().error(message, e);
+								synchronized (mgrToUpdate) {
+									mgrToUpdate.add(mgr1);
+								}
+							}
 						}
 					});
 				}

--- a/bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/menus/DynamicMenuContributionItem.java
+++ b/bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/menus/DynamicMenuContributionItem.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2008, 2015 IBM Corporation and others.
+ * Copyright (c) 2008, 2023 IBM Corporation and others.
  *
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
@@ -10,11 +10,13 @@
  *
  * Contributors:
  *     IBM Corporation - initial API and implementation
+ *     Christoph Läubrich - handle runtime exceptions produced by dynamic contribution
  ******************************************************************************/
 
 package org.eclipse.ui.internal.menus;
 
 import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.ILog;
 import org.eclipse.jface.action.ContributionItem;
 import org.eclipse.jface.action.IContributionItem;
 import org.eclipse.jface.action.IContributionManager;
@@ -62,7 +64,11 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public boolean isDynamic() {
 		if (loadedDynamicContribution != null) {
-			return loadedDynamicContribution.isDynamic();
+			try {
+				return loadedDynamicContribution.isDynamic();
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 		return true;
 	}
@@ -70,7 +76,11 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public boolean isDirty() {
 		if (loadedDynamicContribution != null) {
-			return loadedDynamicContribution.isDirty();
+			try {
+				return loadedDynamicContribution.isDirty();
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 		return super.isDirty();
 	}
@@ -78,7 +88,11 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public boolean isEnabled() {
 		if (loadedDynamicContribution != null) {
-			return loadedDynamicContribution.isEnabled();
+			try {
+				return loadedDynamicContribution.isEnabled();
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 		return super.isEnabled();
 	}
@@ -86,7 +100,11 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public boolean isGroupMarker() {
 		if (loadedDynamicContribution != null) {
-			return loadedDynamicContribution.isGroupMarker();
+			try {
+				return loadedDynamicContribution.isGroupMarker();
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 		return super.isGroupMarker();
 	}
@@ -94,7 +112,11 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public boolean isSeparator() {
 		if (loadedDynamicContribution != null) {
-			return loadedDynamicContribution.isSeparator();
+			try {
+				return loadedDynamicContribution.isSeparator();
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 		return super.isSeparator();
 	}
@@ -102,7 +124,11 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public boolean isVisible() {
 		if (loadedDynamicContribution != null) {
-			return loadedDynamicContribution.isVisible();
+			try {
+				return loadedDynamicContribution.isVisible();
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 		return super.isVisible();
 	}
@@ -110,7 +136,11 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public void saveWidgetState() {
 		if (loadedDynamicContribution != null) {
-			loadedDynamicContribution.saveWidgetState();
+			try {
+				loadedDynamicContribution.saveWidgetState();
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 		super.saveWidgetState();
 	}
@@ -118,7 +148,11 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public void setVisible(boolean visible) {
 		if (loadedDynamicContribution != null) {
-			loadedDynamicContribution.setVisible(visible);
+			try {
+				loadedDynamicContribution.setVisible(visible);
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 		super.setVisible(visible);
 	}
@@ -126,29 +160,49 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public void fill(Composite parent) {
 		IContributionItem contributionItem = getContributionItem();
-		if (contributionItem != null)
-			contributionItem.fill(parent);
+		if (contributionItem != null) {
+			try {
+				contributionItem.fill(parent);
+			} catch (RuntimeException e) {
+				reportErrorForContribution(contributionItem, e);
+			}
+		}
 	}
 
 	@Override
 	public void fill(CoolBar parent, int index) {
 		IContributionItem contributionItem = getContributionItem();
-		if (contributionItem != null)
-			contributionItem.fill(parent, index);
+		if (contributionItem != null){
+			try {
+				contributionItem.fill(parent, index);
+			} catch (RuntimeException e) {
+				reportErrorForContribution(contributionItem, e);
+			}
+		}
 	}
 
 	@Override
 	public void fill(Menu menu, int index) {
 		IContributionItem contributionItem = getContributionItem();
-		if (contributionItem != null)
-			contributionItem.fill(menu, index);
+		if (contributionItem != null) {
+			try {
+				contributionItem.fill(menu, index);
+			} catch (RuntimeException e) {
+				reportErrorForContribution(contributionItem, e);
+			}
+		}
 	}
 
 	@Override
 	public void fill(ToolBar parent, int index) {
 		IContributionItem contributionItem = getContributionItem();
-		if (contributionItem != null)
-			contributionItem.fill(parent, index);
+		if (contributionItem != null) {
+			try {
+				contributionItem.fill(parent, index);
+			} catch (RuntimeException e) {
+				reportErrorForContribution(contributionItem, e);
+			}
+		}
 	}
 
 	private IContributionItem getContributionItem() {
@@ -177,7 +231,11 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public void dispose() {
 		if (loadedDynamicContribution != null) {
-			loadedDynamicContribution.dispose();
+			try {
+				loadedDynamicContribution.dispose();
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 			loadedDynamicContribution = null;
 		}
 		super.dispose();
@@ -186,14 +244,22 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	@Override
 	public void update() {
 		if (loadedDynamicContribution != null) {
-			loadedDynamicContribution.update();
+			try {
+				loadedDynamicContribution.update();
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 	}
 
 	@Override
 	public void update(String id) {
 		if (loadedDynamicContribution != null) {
-			loadedDynamicContribution.update(id);
+			try {
+				loadedDynamicContribution.update(id);
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
 	}
 
@@ -201,8 +267,18 @@ public class DynamicMenuContributionItem extends ContributionItem {
 	public void setParent(IContributionManager parent) {
 		super.setParent(parent);
 		if (loadedDynamicContribution != null) {
-			loadedDynamicContribution.setParent(parent);
+			try {
+				loadedDynamicContribution.setParent(parent);
+			} catch (RuntimeException e) {
+				reportErrorForContribution(loadedDynamicContribution, e);
+			}
 		}
+	}
+
+	private static void reportErrorForContribution(IContributionItem contributionItem, RuntimeException e) {
+		String message = String.format("Dynamic menu contribution '%s' threw an unexpected exception", //$NON-NLS-1$
+				contributionItem);
+		ILog.get().error(message, e);
 	}
 
 }

